### ISSUE TITLE
ppm crashing slapd when cracklib is enabled on RedHat

### ISSUE THREAD

**Issue by JoeCheez** [ltb-project/ppm#7]
I'm experiencing a problem where slapd is crashing when I have cracklib enabled in ppm.  I'm using the yum package below and I built ppm using openldap-2.4.44 and cracklib-2.9.0.

cracklib.x86_64                      2.9.0-11.el7            @anaconda/7.3
cracklib-dicts.x86_64                2.9.0-11.el7            @anaconda/7.3

I suspect that maybe its crashing because the cracklib version i built with might not perfectly match up with the RedHat installed libraries, but its just a theory.  Have you run into this problem before? Any ideas?

**Comment by coudot** [comment]
Did you use LTB ppm RPM or did you compile ppm yourself?

**Comment by JoeCheez** [comment]
I compiled it myself.  I didn't know an rpm existed for it until this morning.  The last link on this page? https://ltb-project.org/rpm/7/x86_64/

**Comment by coudot** [comment]
Yes, ppm is shipped as DEB or RPM with OpenLDAP LTB packages. But I think you need the OpenLDAP LTB version to use the ppm RPM.

**Comment by JoeCheez** [comment]
I installed openldap-ltb and openldap-ltb-ppm and copied/modified my slapd.conf to operate in the new location.  I get the same error, so I definitely think it's a problem with whatever default RedHat cracklib version I'm using.  Is that also packaged with ltb somewhere or is there a specific version I should be using?  The location listed in the ppm.conf downloaded with ltb didn't exist when I looked.

Edit: I should clarify, I don't actually get an error, the last item in the log is 
`ppm: 1 point granted for class class-special`
and the response to the ldappasswd is (-1) Cannot contact LDAP Server and the process is gone.

**Comment by coudot** [comment]
To be honnest, I did not test the cracklib feature on RedHat, it is possible that it is not working at all.

You can check PR #6 to see if it solves the problem.

**Comment by davidcoutadeur** [comment]
I have tested the cracklib feature basically.
Can you give more information on the bug ? Have you tried the test program?
Could you also chek where the cracklib library *and* dictionnaries are? Dictionnaries are the files with extensions .hwm .pwd and .pwi.

**Comment by JoeCheez** [comment]
Setting the log level to "any" doesn't give me any more information.  If  you have any suggestions on how to get more detailed errors anywhere I'd be happy to try it.  What test program are you referring to? `echo Or25@001 | cracklib-check` returns OK.  Here's the results from locate cracklib (minus locale, doc, and man locations)
/usr/lib64/cracklib_dict.hwm
/usr/lib64/cracklib_dict.pwd
/usr/lib64/cracklib_dict.pwi
/usr/lib64/security/pam_cracklib.so
/usr/sbin/cracklib-check
/usr/sbin/cracklib-format
/usr/sbin/cracklib-packer
/usr/sbin/cracklib-unpacker
/usr/sbin/create-cracklib-dict
/usr/share/cracklib
/usr/share/cracklib/cracklib-small.hwm
/usr/share/cracklib/cracklib-small.pwd
/usr/share/cracklib/cracklib-small.pwi
/usr/share/cracklib/cracklib.magic
/usr/share/cracklib/pw_dict.hwm
/usr/share/cracklib/pw_dict.pwd
/usr/share/cracklib/pw_dict.pwi

I don't know why there are 3 different dictionaries.  I've tried them all with the same result.  I've also tried creating a one-value dictionary with create-cracklib-dict just in case there was some kind of bad value in the dictionary.

And just to get one more idiot-check out of the way, I'm not including any file extension on the config line. `cracklibDict /usr/share/cracklib/cracklib-small`

Edit: If you were referring to ppm_test and not cracklib above, this is what I get:
[user@host lib64]$ pwd
/usr/local/openldap/lib64
[user@host lib64]$ ll
total 4972
lrwxrwxrwx 1 ldap ldap      21 Jul 19 09:30 liblber-2.4.so.2 -> liblber-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  208464 Jun  5 16:11 liblber-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  102566 Jun  5 16:11 liblber.a
-rw-r--r-- 1 ldap ldap     866 Jun  5 16:11 liblber.la
lrwxrwxrwx 1 ldap ldap      21 Jul 19 09:30 liblber.so -> liblber-2.4.so.2.10.8
lrwxrwxrwx 1 ldap ldap      21 Jul 19 09:30 libldap-2.4.so.2 -> libldap-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap 1334624 Jun  5 16:11 libldap-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  550608 Jun  5 16:11 libldap.a
-rw-r--r-- 1 ldap ldap     926 Jun  5 16:11 libldap.la
lrwxrwxrwx 1 ldap ldap      23 Jul 19 09:30 libldap_r-2.4.so.2 -> libldap_r-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap 1450320 Jun  5 16:11 libldap_r-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  610202 Jun  5 16:11 libldap_r.a
-rw-r--r-- 1 ldap ldap     949 Jun  5 16:11 libldap_r.la
lrwxrwxrwx 1 ldap ldap      23 Jul 19 09:30 libldap_r.so -> libldap_r-2.4.so.2.10.8
lrwxrwxrwx 1 ldap ldap      21 Jul 19 09:30 libldap.so -> libldap-2.4.so.2.10.8
lrwxrwxrwx 1 ldap ldap      22 Jul 19 09:30 libslapi-2.4.so.2 -> libslapi-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  519488 Jun  5 16:11 libslapi-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap  207158 Jun  5 16:11 libslapi.a
-rw-r--r-- 1 ldap ldap     864 Jun  5 16:11 libslapi.la
lrwxrwxrwx 1 ldap ldap      22 Jul 19 09:30 libslapi.so -> libslapi-2.4.so.2.10.8
-rw-r--r-- 1 ldap ldap   68808 Jun  5 16:11 ppm.so
-rwxr-xr-x 1 ldap ldap    7480 Jun  5 16:11 ppm_test
[user@host lib64]$ ./ppm_test
./ppm_test: error while loading shared libraries: ppm.so: cannot open shared object file: No such file or directory

**Comment by davidcoutadeur** [comment]
Yes, I talked about ppm_test.
As mentionned in README, you can use the test program like: 
```
cd /usr/local/openldap/lib64
PPM_CONFIG_FILE=/usr/local/openldap/etc/openldap/ppm.conf LD_LIBRARY_PATH=. ./ppm_test "my_password" && echo OK
```
(non working password with default ppm.conf)
or
```
cd /usr/local/openldap/lib64
PPM_CONFIG_FILE=/usr/local/openldap/etc/openldap/ppm.conf LD_LIBRARY_PATH=. ./ppm_test "my_password2" && echo OK
```
(working password with default ppm.conf)

**Comment by davidcoutadeur** [comment]
Bug reproduced and confirmed on Debian 8.
This is due to a linking problem with cracklib, as brilliantly reported in #6 
You could try compiling again with this line modified: (just add the CRACK_LIB variable to the end)

```
ppm: clean ppm.o
        $(CC) $(LDAP_INC) -shared -o ppm.so ppm.o $(CRACK_LIB)
```

A fix will be proposed shortly.

**Comment by davidcoutadeur** [comment]
Last trunk version is now fixing the problem.
@JoeCheez Can you test all is ok for you?

**Comment by JoeCheez** [comment]
The ppm make was unable to find lcrack after adding $(CRACK_LIB).  I ended up downloading the cracklib-2.9.0 source and installing that.  Then, the ppm make was able to find everything after manually including the openldap source location.

make CONFIG=/app/openldap/etc/openldap/ppm.conf LIBDIR=/app/openldap/lib LDAP_INC="-I/app/openldap-2.4.44/include -I/app/openldap-2.4.44/servers/slapd"

After copying ppm.so over to my ldap install and setting all the quality checks to 0 I was able to trigger cracklib successfully without OpenLDAP crashing.  Thanks for all your help.
